registry-v2.md

```diff
@@ -0,0 +1,82 @@
+# Registry v2
+
+*NOTE: This registry is not yet deployed on hex.pm*
+
+The first version of the registry format used a single file for the entire registry. While this was efficient in the number of HTTP requests needed to resolve dependencies, the size of the registry would grow out of control as the repository grows.
+
+The new version of the registry format is split into multiple files to make it scale as the repository grows with more packages.
+
+For every package in the repository there is a package registry file under `/packages/[NAME]` in the repository, it contains all the releases of that package and all dependencies of the releases. There are also two index files `/names` and `/versions` that contains all package names and all package names+versions respectively.
+
```

Comment:
We have the index resources `/names` and `/versions` for discovery. They have the caveat that they should generally be avoided when possible since they will only grow larger and larger in size forever.

The mix client will never use the index resources since we fetch from the registry on every resolution. Rebar3 has an explicit `rebar3 update` step so you can use them if you want. The reason is that users expect resolution to be quick but they are usually fine if an explicit update command takes a few seconds.

We can also consider supporting http range requests on `/names` and `/versions` to make subsequent requests more efficient. This is tricky (but not impossible) when they are encoded with protobufs because it requires knowledge about the internal protobuf encoding. Another downside is that it's much trickier and less efficient to compress. We would also need to change the format to reflect changes to the registry instead of a snapshot of the registry. An examples of the `/names` resource would be:

```
ADD ecto
ADD plug
REMOVE poison
ADD recon
```